### PR TITLE
LPS-141227 & LPS-141228 Test automation for Remote App - Custom element fields can be created & edited

### PR DESCRIPTION
**Pull request by Esther-OC**
**Description**
Two new Poshi tests to verify Remote Apps:
- CustomElementFieldsCanBeSaved --> it creates a new Custom Element where several specific fields of the form can be filled in and saved.
- CustomElementCanBeEdited --> it creates a Custom Element and then edits and saves the content of the fields.

**Test Plan**

- [x]  make sure test passes on CI.
 
**JIRA Ticket**
- Test CustomElementFieldsCanBeSaved: https://issues.liferay.com/browse/LPS-141227
- Test CustomElementCanBeEdited:  https://issues.liferay.com/browse/LPS-141228

### CODE REVIEW

**Comment by Esther-OC** [comment]
ci:test:sf

**Comment by john-co** [comment]
Please commit and push the SF fix.

**Comment by john-co** [comment]
ci:test:frontend:nocompile

File: portal-web/test/functional/com/liferay/portalweb/macros/RemoteAppsEntry.macro [hunk 1]
```diff
@@ -12,21 +12,44 @@ definition {
 			value1 = "${entryHTMLName}");
 
 		Type(
-			key_id = "customElementURLs",
-			locator1 = "RemoteAppsEntry#URL",
+			locator1 = "RemoteAppsEntry#CUSTOM_ELEMENT_URL",
```
Comment (john-co) [inline portal-web/test/functional/com/liferay/portalweb/macros/RemoteAppsEntry.macro:15]: why was the RemoteAppsEntry#URL refactored this way?

File: portal-web/test/functional/com/liferay/portalweb/tests/enduser/frontendinfrastructure/uiinfrastructure/remotewebapps/RemoteApps.testcase [hunk 2]
```diff
@@ -22,6 +22,22 @@ definition {
 		}
 	}
 
+	@description = "Verify remote app can be created"
+	@priority = "5"
+	test CanBeCreated {
```
Comment (john-co) [inline portal-web/test/functional/com/liferay/portalweb/tests/enduser/frontendinfrastructure/uiinfrastructure/remotewebapps/RemoteApps.testcase:27]: This is from IFrameCanBeCreated right?

File: portal-web/test/functional/com/liferay/portalweb/tests/enduser/frontendinfrastructure/uiinfrastructure/remotewebapps/RemoteApps.testcase [hunk 3]
```diff
@@ -30,8 +46,6 @@ definition {
 
 		var remoteAppName = "My New Remote App To Delete";
 
-		RemoteApps.tearDown();
-
```
Comment (john-co) [inline portal-web/test/functional/com/liferay/portalweb/tests/enduser/frontendinfrastructure/uiinfrastructure/remotewebapps/RemoteApps.testcase:34]: I think you accidentally left this out.

File: portal-web/test/functional/com/liferay/portalweb/tests/enduser/frontendinfrastructure/uiinfrastructure/remotewebapps/RemoteApps.testcase [hunk 4]
```diff
@@ -65,63 +79,82 @@ definition {
 			entryType = "${customElementType}");
 	}
 
-	@description = "Verify that remote app of type Custom Element can be displayed by portlet"
-	@priority = "5"
-	test CustomElementCanBeDisplayedByPortlet {
```
Comment (john-co) [inline portal-web/test/functional/com/liferay/portalweb/tests/enduser/frontendinfrastructure/uiinfrastructure/remotewebapps/RemoteApps.testcase:70]: This existing testcase was deleted, please add back...

File: portal-web/test/functional/com/liferay/portalweb/tests/enduser/frontendinfrastructure/uiinfrastructure/remotewebapps/RemoteApps.testcase [hunk 5]
```diff
@@ -65,63 +79,82 @@ definition {
...
 				entryURL = "${customElementURL}");
-		}
 
-		task ("Add Vanilla Counter to Test Page") {
-			ContentPagesNavigator.openEditContentPage(
-				pageName = "Test Page",
-				siteName = "Guest");
+		Click(
+				key_tableEntryName = "${customElementName}",
+				locator1 = "RemoteApps#TABLE_ENTRY_NAME_REMOTE_TABLE");
```
Comment (john-co) [inline portal-web/test/functional/com/liferay/portalweb/tests/enduser/frontendinfrastructure/uiinfrastructure/remotewebapps/RemoteApps.testcase:104]: tabs vs spacing, check throughout the test.

File: portal-web/tests/css/base-style.css [hunk 6]
```diff
@@ -0,0 +1,179 @@
+
+body {
+    margin: 0;
+    padding: 0;
+    font-family: sans-serif;
+    font-size: 12pt;
```
Comment (john-co) [inline portal-web/tests/css/base-style.css:6]: Please remove extra files that are not related.

File: portal-web/tests/css/style.css [hunk 7]
```diff
@@ -0,0 +1,84 @@
+
+#summary {
+    margin-top: 30px;
+    margin-bottom: 40px;
+}
```
Comment (john-co) [inline portal-web/tests/css/style.css:5]: not needed

File: portal-web/tests/js/report.js [hunk 8]
```diff
@@ -0,0 +1,194 @@
+(function (window, document) {
+    "use strict";
+
+    var tabs = {};
+
+    function changeElementClass(element, classValue) {
```
Comment (john-co) [inline portal-web/tests/js/report.js:6]: not related

**Review by john-co — CHANGES_REQUESTED** [review]
Please make the requested changes and resubmit.

File: portal-web/test/functional/com/liferay/portalweb/macros/RemoteApps.macro [hunk 9]
```diff
@@ -34,6 +35,66 @@ definition {
...
+		Click(locator1 = "RemoteAppsEntry#REMOTE_APPS_INSTANCEABLE");
+
+		PortletEntry.publish();
+	}
+
+	macro editCustomElement {
+
+		TripleClick(
+				key_text = "Name",
+				locator1 = "TextInput#ANY");
```
[the remark this answers is not in the document]
Comment (Esther-OC) [reply portal-web/test/functional/com/liferay/portalweb/macros/RemoteApps.macro]: Type cannot be used to edit the field of the form because it just add the new text to the previous one.
So triple click is the solution I found to select the previous text and overwrite it typing the new one.  (Triple click because it can select all the text, even if there are spaces between words).

File: portal-web/test/functional/com/liferay/portalweb/macros/RemoteAppsEntry.macro (hunk 1 above)
Comment (Esther-OC) [reply portal-web/test/functional/com/liferay/portalweb/macros/RemoteAppsEntry.macro:15]: My mistake.  The problem with the URLs was solved before my changes and it seems I didn't realize and made a mistake when comparing and merging the changes before committing them.

File: portal-web/test/functional/com/liferay/portalweb/tests/enduser/frontendinfrastructure/uiinfrastructure/remotewebapps/RemoteApps.testcase (hunk 2 above)
Comment (Esther-OC) [reply portal-web/test/functional/com/liferay/portalweb/tests/enduser/frontendinfrastructure/uiinfrastructure/remotewebapps/RemoteApps.testcase:27]: Yes.  Same as before, there were some changes in this test case that didn't appear in my conflicts.  I have changed it now.

File: portal-web/test/functional/com/liferay/portalweb/tests/enduser/frontendinfrastructure/uiinfrastructure/remotewebapps/RemoteApps.testcase (hunk 4 above)
Comment (Esther-OC) [reply portal-web/test/functional/com/liferay/portalweb/tests/enduser/frontendinfrastructure/uiinfrastructure/remotewebapps/RemoteApps.testcase]: Done.

File: portal-web/tests/js/report.js (hunk 8 above)
Comment (Esther-OC) [reply portal-web/tests/js/report.js]: Don't understand why these files have been added, they were under the "test" folder, but I don't know if they were added automatically to the commit because I always made "git add file_name" and only wiht the .testcase, .macro and .path files we are working on.

**Comment by Esther-OC** [comment]
The PR has been updated with the changes required and included the changes in @sergiojimcos tests from his PR https://github.com/liferay-frontend/liferay-portal/pull/1719

**Comment by Esther-OC** [comment]
ci:test:sf

**Comment by Esther-OC** [comment]
@john-co  I marked most of your requests as solved, ones because I made the changes and others with my answers.  Is it ok or should you be the one to mark the conversations as resolved?

**Comment by Esther-OC** [comment]
ci:test:sf

**Comment by Esther-OC** [comment]
ci:test:sf

**Comment by john-co** [comment]
ci:test:frontend:nocompile

**Comment by john-co** [comment]
Remember to include a poshi report (index.html) to the JIRA ticket to show the test passes locally and some proof that the tests pass on CI. For new Remote apps tests it will be sufficient to run `ci:test:frontend:nocompile`. This will make it easier for reviewer to check these.

**Comment by john-co** [comment]
> I marked most of your requests as solved, ones because I made the changes and others with my answers. Is it ok or should you be the one to mark the conversations as resolved?

Ok to mark the ones fixed as resolved. For the ones that you only responded with no changes, leave them unresolved in case reviewer has additional feedback/questions.

File: portal-web/test/functional/com/liferay/portalweb/macros/RemoteApps.macro [hunk 10]
```diff
@@ -34,6 +34,64 @@ definition {
...
+			key_text = "CSS URL",
+			locator1 = "TextInput#ANY",
+			value1 = "${entryCSSURL}");
+
+		AssertTextEquals(
+			custom_entryProperties = "${entryProperties}",
+			key_text = "Properties",
+			locator1 = "RemoteAppsEntry#REMOTE_APPS_PROPERTIES",
+			value1 = "${entryProperties}");
+	}
```
Comment (john-co) [inline portal-web/test/functional/com/liferay/portalweb/macros/RemoteApps.macro:86]: Probably fine to have all these in one macro. May eventually need to refactor if we have another test that asserts other fields.

File: portal-web/test/functional/com/liferay/portalweb/tests/enduser/frontendinfrastructure/uiinfrastructure/remotewebapps/RemoteApps.testcase [hunk 11]
```diff
@@ -106,6 +129,83 @@ definition {
...
+		var newCustomElementTMLName = "vanilla-counter-edited";
+		var newCustomElementCSSURL = "http://remote-component-test.wincent.com/index.css";
+		var newCustomElementProperties = "test-data-user=QAuser";
+
+		RemoteApps.goToRemoteAppsPortlet();
+
+		RemoteApps.addCustomElement(
+			entryHTMLName = "${customElementHTMLName}",
+			entryName = "${customElementName}",
+			entryURL = "${customElementURL}");
```
Comment (john-co) [inline portal-web/test/functional/com/liferay/portalweb/tests/enduser/frontendinfrastructure/uiinfrastructure/remotewebapps/RemoteApps.testcase:150]: Use JSON for any setup/given steps.

File: portal-web/test/functional/com/liferay/portalweb/macros/RemoteApps.macro [hunk 12]
```diff
@@ -49,7 +107,47 @@ definition {
...
+
+	macro editCustomElement {
+		TripleClick(
+			key_text = "Name",
+			locator1 = "TextInput#ANY");
+
+		Type(
+			key_text = "Name",
+			locator1 = "TextInput#ANY",
+			value1 = "${entryName}");
```
[the remark this answers is not in the document]
Comment (john-co) [reply portal-web/test/functional/com/liferay/portalweb/macros/RemoteApps.macro:121]: I ran the test locally without the TripleClicks and it worked for me. I think the thing about the Type.function is that it clears the input field by default already.

**Comment by john-co** [comment]
I went ahead and forwarded @sergiojimcos 's PR in https://github.com/liferay-frontend/liferay-portal/pull/1719 since that part looked good.

Please make the requested changes, and clean up the commit history. Some commits should be squashed/combined to remove unnecessary history such as https://github.com/liferay-frontend/liferay-portal/pull/1728/commits/59935eeb8613ae00ec68550af4e99b0c7acc6dc5

**Comment by john-co** [comment]
I removed some extra files that were included in last commit and squashed the first two commits since there was extra noise that we don't want to commit.

Please continue to refactor the additional changes.

File: portal-web/test/functional/com/liferay/portalweb/tests/enduser/frontendinfrastructure/uiinfrastructure/remotewebapps/RemoteApps.testcase [hunk 13]
```diff
@@ -106,6 +129,83 @@ definition {
...
+			locator1 = "RemoteApps#TABLE_ENTRY_NAME_REMOTE_TABLE");
+
+		RemoteApps.assertCustomElementFields(
+			entryCSSURL = "${customElementCSSURL}",
+			entryHTMLName = "${customElementHTMLName}",
+			entryName = "${customElementName}",
+			entryProperties = "${customElementProperties}",
+			entryURL = "${customElementURL}");
+	}
+
```
[the remark this answers is not in the document]
Comment (Esther-OC) [reply portal-web/test/functional/com/liferay/portalweb/tests/enduser/frontendinfrastructure/uiinfrastructure/remotewebapps/RemoteApps.testcase:208]: Test steps wrapped in tasks for the following test cases:
- CustomElementCanBeDeleted
- CustomElementCanBeEdited
- CustomElementFieldsCanBeSaved

File: portal-web/test/functional/com/liferay/portalweb/macros/RemoteApps.macro (hunk 12 above)
[the remark this answers is not in the document]
Comment (Esther-OC) [reply portal-web/test/functional/com/liferay/portalweb/macros/RemoteApps.macro:121]: OMG, it works just with the Type.function!   I run the test locally and I really tried it several times at the beginning when writing the test and it didn't work (it just added the text to the previous one).  It's obvious that something else was different and this is why I figured out the triple click solution.
I feel quite frustrated and dumb now  🥴  😭

File: portal-web/test/functional/com/liferay/portalweb/tests/enduser/frontendinfrastructure/uiinfrastructure/remotewebapps/RemoteApps.testcase [hunk 14]
```diff
@@ -106,6 +129,83 @@ definition {
...
+	test CustomElementCanBeEdited {
+		property portal.acceptance = "true";
+
+		var customElementName = "New Counter";
+		var customElementHTMLName = "new-counter";
+		var customElementURL = "http://remote-component-test.wincent.com/packages/vanilla-counter/index.js";
+		var newCustomElementName = "Vanilla Counter Edited";
+		var newCustomElementTMLName = "vanilla-counter-edited";
+		var newCustomElementCSSURL = "http://remote-component-test.wincent.com/index.css";
+		var newCustomElementProperties = "test-data-user=QAuser";
```
[the remark this answers is not in the document]
Comment (Esther-OC) [reply portal-web/test/functional/com/liferay/portalweb/tests/enduser/frontendinfrastructure/uiinfrastructure/remotewebapps/RemoteApps.testcase]: Done!  Variables for the setup/given steps of the following test cases are now inside a JSON macro:
- CustomElementCanBeCreated
- CustomElementCanBeEdited
- CustomElementFieldsCanBeSaved

**Comment by Esther-OC** [comment]
ci:test:frontend:nocompile

**Comment by Esther-OC** [comment]
Continue in https://github.com/liferay-frontend/liferay-portal/pull/1764

**Comment by Esther-OC** [comment]
ci:test:frontend:nocompile

File: portal-web/test/functional/com/liferay/portalweb/macros/RemoteApps.macro (hunk 12 above)
[the remark this answers is not in the document]
Comment (john-co) [reply portal-web/test/functional/com/liferay/portalweb/macros/RemoteApps.macro]: It's ok, now you know how the Type function should work. There's some hidden things within the poshi functions that aren't that clear unless we dive deeper into what they really do with selenium.

Other weird things with finicky computers could have happened too.

**Comment by john-co** [comment]
Much better! Looks good from the overall review. There's just a couple Remote app tests failing, could you double check on those? Once the tests are resolved this looks good to forward.

**Comment by Esther-OC** [comment]
ci:test:frontend:nocompile

**Comment by Esther-OC** [comment]
ci:test:frontend:nocompile

**Comment by Esther-OC** [comment]
ci:test:frontend:nocompile

**Comment by Esther-OC** [comment]
ci:forward

**Comment by magjed4289** [comment]
ci:test:sf

**Comment by magjed4289** [comment]
Just started reviewing :)

**Comment by Esther-OC** [comment]
ci:forward
